cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp

```diff
@@ -0,0 +1,272 @@
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl.hpp"
+#include "oneapi/dal/backend/common.hpp"
+#include "oneapi/dal/detail/common.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/detail/policy.hpp"
+#include "oneapi/dal/detail/profiler.hpp"
+#include "oneapi/dal/backend/memory.hpp"
+#include "oneapi/dal/algo/objective_function/backend/objective_impl.hpp"
+#include "oneapi/dal/backend/primitives/objective_function.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+namespace oneapi::dal::objective_function::backend {
+
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+using alloc = sycl::usm::alloc;
+
+using bk::context_gpu;
+using task_t = task::compute;
+using input_t = compute_input<task_t>;
+using result_t = compute_result<task_t>;
+
+std::int64_t get_block_size(std::int64_t n, std::int64_t p) {
+    constexpr std::int64_t max_alloc_size = 1 << 21;
+    return p > max_alloc_size ? 512 : max_alloc_size / p;
+}
+
+template <typename Float>
+result_t compute_kernel_dense_batch_impl<Float>::operator()(
+    const detail::descriptor_base<task_t>& desc,
+    const input_t& input) {
+    ONEDAL_ASSERT(input.get_data().has_data());
+    ONEDAL_ASSERT(input.get_parameters().has_data());
+    ONEDAL_ASSERT(input.get_responses().has_data());
+    const auto data = input.get_data();
+    const auto params = input.get_parameters();
+    const auto responses = input.get_responses();
+    const std::int64_t n = data.get_row_count();
+    const std::int64_t p = data.get_column_count();
+    ONEDAL_ASSERT(responses.get_row_count() == n);
+    ONEDAL_ASSERT(responses.get_column_count() == 1);
+    ONEDAL_ASSERT(params.get_row_count() == p + 1);
+    ONEDAL_ASSERT(params.get_column_count() == 1);
+
+    auto obj_impl = detail::get_objective_impl(desc);
+
+    const Float L1 = obj_impl->get_l1_regularization_coefficient();
+    const Float L2 = obj_impl->get_l2_regularization_coefficient();
+
+    const std::int64_t bsz = get_block_size(n, p);
+
+    const auto params_nd = pr::table2ndarray_1d<Float>(q_, params, alloc::device);
+    const auto* const params_ptr = params_nd.get_data();
+
+    const auto responses_nd_big = pr::table2ndarray_1d<std::int32_t>(q_, responses, alloc::device);
+
+    auto result = compute_result<task_t>{}.set_result_options(desc.get_result_options());
+
+    auto [out, out_e] = pr::ndarray<Float, 1>::zeros(q_, { p + 2 }, sycl::usm::alloc::device);
+    out_e.wait_and_throw();
```

Comment:
Too many calls of `.wait_and_throw()`. Please review it and pass events correctly